core/server/data/migrations/versions/5.0/2022-04-21-15-44-backfill-mrr-events-for-canceled-subscriptions.js

```diff
@@ -0,0 +1,66 @@
+
+const {createTransactionalMigration} = require('../../utils');
+
+module.exports = createTransactionalMigration(
+    async function up(knex) {
+        const subscriptions = await knex('members_stripe_customers_subscriptions')
+            .leftJoin('members_paid_subscription_events', 'members_stripe_customers_subscriptions.id', '=', 'members_paid_subscription_events.subscription_id')
+            .join('members_stripe_customers', 'members_stripe_customers_subscriptions.customer_id', '=', 'members_stripe_customers.customer_id')
+            .join('members', 'members_stripe_customers.member_id', '=', 'members.id')
+            .where('members_stripe_customers_subscriptions.cancel_at_period_end', '=', true)
+            .where('members_stripe_customers_subscriptions.mrr', '!=', 0)
+            .where(knex.raw(`(members_paid_subscription_events.type = 'canceled' OR members_paid_subscription_events.type is null)`))
+            .select(
+                'members_paid_subscription_events.id AS event_id',
+                'members_stripe_customers_subscriptions.id',
+                'members_stripe_customers_subscriptions.updated_at',
+                'members_stripe_customers_subscriptions.mrr',
+                'members_stripe_customers_subscriptions.stripe_price_id',
+                'members_stripe_customers_subscriptions.plan_currency',
+                'members_stripe_customers.member_id'
+            );
+
+        if (subscriptions.length === 0) {
+            logging.info('No canceled subscriptions found, skipping migration.');
+            return;
+        } else {
+            logging.info(`Found ${subscriptions.length} canceled subscriptions, inserting MRR events.`);
+        }
+
+        const {toUpdate, toInsert} = subscriptions.reduce((memo, subscription) => {
+            if (subscription.event_id) {
+                return {
+                    toInsert: memo.toInsert,
+                    toUpdate: memo.toUpdate.concat({
+                        id: subscription.event_id,
+                        mrr_delta: -subscription.mrr
+                    })
+                };
+            }
+
+            return {
+                toUpdate: memo.toUpdate,
+                toInsert: memo.toInsert.concat({
+                    id: ObjectID().toHexString(),
+                    type: 'canceled',
+                    source: 'migration',
+                    created_at: subscription.updated_at,
+                    from_plan: subscription.stripe_price_id,
+                    to_plan: subscription.stripe_price_id,
+                    subscription_id: subscription.id,
+                    member_id: subscription.member_id,
+                    currency: subscription.plan_currency,
+                    mrr_delta: -subscription.mrr
+                })
+            };
+        }, {toUpdate: [], toInsert: []});
+
+        await knex.batchInsert('members_paid_subscription_events', toInsert);
+        for (const event of toUpdate) {
```

Comment:
Not sure about this, I think there's probably some nested update query we can do - not sure if it's worth the complexity though